Rename trio.hazmat → trio.lowlevel

njsmith opened the pull request:
Closes: gh-476

pquentin commented on newsfragments/476.removal.rst:
```diff
@@ -0,0 +1,3 @@
+It turns out that everyone got confused by the name ``trio.lowlevel``.
```
s/lowlevel/hazmat/s

njsmith replied:
Haha, whoops! nice catch :-)

georgepsarakis commented on docs/source/reference-lowlevel.rst:
```diff
@@ -1,44 +1,40 @@
+You don't need to be scared of :mod:`trio.lowlevel`, as long as you
+take proper precautions. These are real public APIs, with strictly
+defined and carefully documented semantics. But, you should read
+carefully and take proper precautions. Some of those strict semantics
```
Is the phrase repetition with line 28 intentional? Perhaps something like this could also work:

```
But, you should pay attention while using them and certainly read the documentation in detail.
```

njsmith replied:
Good point, tweaked the wording a bit.

njsmith commented:
@pquentin 
> This no longer works, is it expected?
> ```
> >>> import trio.hazmat
> ```

Excellent question! I did not think about that.

I doubt many folks are writing `import trio.hazmat`, because `import trio` automatically gives you `trio.hazmat`. But I bet there are folks writing `from trio.hazmat import ...`, and it turns out that this change also broke that, for the same reason.

The issue is that when you refer to `trio.hazmat.something`, then it looks for the `hazmat` attribute on `trio`, and when the attribute isn't found, then it runs our `__getattr__` fallback code, which issues the warning and then returns the proper value. BUT, `import trio.hazmat` tries to find `trio/hazmat.py` directly, without checking for an existing `.hazmat` attribute on `trio`.

I did find a workaround though: if we make a `trio/hazmat.py` file that issues a warning and then does `from trio.lowlevel import *`, then everything mostly works:

- If you do `trio.hazmat.blah` then it works and gives a warning
- If you do `from trio.hazmat import blah` then it works and gives a warning

The one issue is that *after* you do `from trio.hazmat import ...`, then any *future* accesses to `trio.hazmat.whatever` will stop issuing warnings (because the import machinery adds a real `hazmat` attribute to the `trio` module, so our `__getattr__` fallback code never runs). But that's OK... it's technically not *quite* as nice because it defeats the part of the `warnings` module that tries to tell you about all the different places that are using the deprecated code, but we still preserve the key invariant that every program which uses `trio.hazmat` does print at least one warning.

njsmith commented:
Addressed review comments.

On further consideration, I think I want to make one more change: in the changelog, it's a bit awkward to handle all the old references to `trio.hazmat`, since on the one hand it's a historical document so you want to leave things alone, but on the other hand, it's something people read in the present.

The awkward compromise I did in the initial commit was to apply the search/replace to only the text that had live links. I think a better way would be to keep the text the same while fixing the link destinations only.

njsmith commented:
OK done, I think (hope?) this is ready now

oremanj commented:
Looks great, thank you!

pquentin commented:
Looks good to me too, thanks!
